src/contours.rs

```diff
@@ -0,0 +1,193 @@
+//! Functions for detecting contours of polygons in an image and approximating
+//! polygon from set of points.
+
+use crate::definitions::Point;
+use image::GrayImage;
+use num::{cast, Num, NumCast};
+use std::collections::VecDeque;
+
+/// Contour struct containing its points, is_outer flag to determine whether the
+/// contour is an outer border or hole border, and the parent option (the index
+/// of the parent contour in the contours vec)
+#[derive(Debug)]
+pub struct Contour<T: Num + NumCast + Copy + PartialEq + Eq> {
+    /// All the points on the contour
+    pub points: Vec<Point<T>>,
+    /// Flag to determine whether the contour is an outer border or hole border
+    pub is_outer: bool,
+    /// the index of the parent contour in the contours vec, or None if the
+    /// contour is the outermost contour in the image
+    pub parent: Option<usize>,
+}
+impl<T: Num + NumCast + Copy + PartialEq + Eq> Contour<T> {
+    /// Construct a contour.
+    pub fn new(points: Vec<Point<T>>, is_outer: bool, parent: Option<usize>) -> Self {
+        Contour {
+            points,
+            is_outer,
+            parent,
+        }
+    }
+}
+
+/// Finds all the points on the contours on the provided image.
+/// Handles all non-zero pixels as 1.
+pub fn find_contours<T: Num + NumCast + Copy + PartialEq + Eq>(
+    original_image: &GrayImage,
+) -> Vec<Contour<T>> {
+    find_contours_with_thresh(original_image, 0)
+}
+
+/// Finds all contours (contour - all the points on the edge of a polygon)
+/// in the provided image. The algorithm works only with a binarized image,
+/// therefore, the `thresh` param defines the value for which every pixel with
+/// value higher then `thresh` will be considered as 1, and 0 otherwise.
+///
+/// Based on the algorithm proposed by Suzuki and Abe: Topological Structural
+/// Analysis of Digitized Binary Images by Border Following
+///
+pub fn find_contours_with_thresh<T: Num + NumCast + Copy + PartialEq + Eq>(
+    original_image: &GrayImage,
+    thresh: u8,
+) -> Vec<Contour<T>> {
+    let width = original_image.width() as usize;
+    let height = original_image.height() as usize;
+    let mut image_values = vec![vec![0i32; height]; width];
+
+    for y in 0..height {
+        for x in 0..width {
+            if original_image.get_pixel(x as u32, y as u32).0[0] > thresh {
+                image_values[x][y] = 1;
+            }
+        }
+    }
+    let mut diffs = VecDeque::from(vec![
+        (-1, 0),  // w
+        (-1, -1), // nw
+        (0, -1),  // n
+        (1, -1),  // ne
+        (1, 0),   // e
+        (1, 1),   // se
+        (0, 1),   // s
+        (-1, 1),  // sw
+    ]);
+    let mut x = 0;
+    let mut y = 0;
+    let last_pixel = (width - 1, height - 1);
+
+    let mut contours: Vec<Contour<T>> = Vec::new();
+    let mut skip_tracing;
+    let mut nbd = 1;
+    let mut lnbd = 1;
```

Comment:
Could you give clearer names to these two values.

Reply:
Sure, I've used the notations from the [paper](https://www.semanticscholar.org/paper/Topological-structural-analysis-of-digitized-binary-Suzuki-Abe/cf021db5e811fd5b67ee3aa4db0a6a0351d276d2), but it makes sense to give them clearer names.